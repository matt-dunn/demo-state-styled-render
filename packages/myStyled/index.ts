/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

export { myStyled as default } from "./myStyled";

export { css, cssWithProps } from "./css";
