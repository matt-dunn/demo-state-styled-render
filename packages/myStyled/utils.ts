/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import Stylis from "stylis";

import { AnyRules, StylesheetPartial } from "./stylesheet";

const CLASS_NAME_PREFIX = "ms__";

const DEBUG = process.env.DEBUG === "true";

function isFunction(value: any): value is (...args: any[]) => any {
  return typeof value === "function";
}

export const parseRule = <T>(
  strings: TemplateStringsArray,
  args?: T[],
  props?: any
): string =>
  strings
    .reduce((rule: string[], part: string, index: number) => {
      rule.push(part);

      const arg = args?.[index];
      const value = arg && isFunction(arg) ? arg(props) : arg;

      if (value) {
        rule.push(value);
      }

      return rule;
    }, [])
    .join("");

// Use a fixed class prefix to simplify client/server class names
export const generateClassName = (hash: string): string =>
  `${CLASS_NAME_PREFIX}${hash}`;

const stylis = new Stylis({
  global: false,
});

export const updateSheetRule = (
  sheet: StylesheetPartial<CSSRuleList | AnyRules>,
  className: string,
  rule: string
) => {
  const DEBUG_RULES: string[] = [];

  const selectorText = `.${className}`;

  // See https://github.com/thysultan/stylis.js#plugins for plugin details
  stylis.use((context, content, selectors, parent) => {
    const selector = selectors[0];
    if (selector) {
      // Remove any additional specificity...
      // TODO: a bit hacky... can likely be improved/made more robust. this is intended to ensure child blocks are not inserted without the parent wrapper (e.g. @media rules)
      const normalizedSelector = `${parent[0] || ""}${
        (parent[0] &&
          selector.toString().replace(new RegExp(parent[0], "g"), "")) ||
        ""
      }`;

      // Do not include any global styles... Should be handled... globally! 2 = selector block, 3 = @at-rule block
      if (
        ((context === 2 && selector.startsWith(selectorText)) ||
          context === 3) &&
        normalizedSelector !== parent[0] &&
        selector.toLowerCase().indexOf(":global") === -1
      ) {
        // Add class postfix to localise animation name if required
        const currentSelector = `${selector}${
          (selector.toLowerCase().startsWith("@keyframes ") &&
            `-${className}`) ||
          ""
        }`;
        // Set the correct content if font-face
        const currentContent =
          (selector.toLowerCase().startsWith("@font-face") &&
            content.substr(1, content.length - 2)) ||
          content;

        sheet.insertRule(`${currentSelector} {${currentContent}}`);

        if (DEBUG) {
          DEBUG_RULES.push(`${currentSelector} {\n    ${content}\n  }`);
        }
      }
    }
  });

  stylis(selectorText, rule);

  if (DEBUG) {
    console.log(
      `UPDATE(${sheet.rules.length} rules)\n `,
      DEBUG_RULES.join("\n  ")
    );
  }

  return className;
};
