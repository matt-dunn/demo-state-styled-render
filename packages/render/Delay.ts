/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {AnyNode, Node} from "./";
import { useEffect, useState } from "./hooks";

type DelayProps = {
  delay?: number;
  children: Node;
};

export const Delay = ({ delay = 500, children }: DelayProps): AnyNode => {
  const [waiting, setWaiting] = useState(delay !== 0);

  useEffect(() => {
    if (waiting) {
      const t = setTimeout(() => {
        setWaiting(false);
      }, delay);

      return () => {
        clearTimeout(t);
      };
    }
  }, [delay]);

  return waiting ? null : children;
};
