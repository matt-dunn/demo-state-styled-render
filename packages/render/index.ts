/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

export { Mx as default } from "./render";

export { Node, Children, FC, HTMLElementMap } from "./types";

export { classnames } from "./utils";

export { jsx, jsxFrag, createElement } from "./element";

export * from "./hooks";
