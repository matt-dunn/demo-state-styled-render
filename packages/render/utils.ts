/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { Node } from "./types";

type MappedProps = {
  [key: string]: string;
};

const mappedProps: MappedProps = {
  className: "class",
  htmlFor: "for",
};

export const propertyMap = (name: string) => {
  if (mappedProps[name]) {
    return mappedProps[name];
  } else if (/^on/.test(name)) {
    return name.toLowerCase();
  }

  return name;
};

export const isNode = (node: any): node is Node =>
  (typeof node === "object" &&
    Object.prototype.hasOwnProperty.call(node, "type")) ||
  false;

export const classnames = (...classNames: (string | undefined)[]): string =>
  classNames.join(" ");

type LooseRef<T = any> = {
  current: T;
};

export const looseRef = <T>(initial: T): LooseRef<T> => ({
  current: initial,
});
