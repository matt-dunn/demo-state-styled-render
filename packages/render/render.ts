/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {Actions, Store} from "packages/state";

import {FC, Node} from "./types";
import {updateTree} from "./element";

type UseState = <S = any>(initialState: S) => [S, (newValue: S) => S];

type MxContainer = {
  index: number;
  state: any[];
  currentState: any;
  currentTree?: Node;
  renderTree: () => void;
  useState: UseState;
  render: <S>(store: Store<S>, actions: Actions) => (component: FC) => (mountPoint: HTMLElement) => void;
}

export const Mx: MxContainer = {
  index: 0,
  state: [],
  currentState: undefined,
  currentTree: undefined,
  renderTree: undefined as unknown as () => void,
  useState: <S = any>(initialState: S) => {
    const activeIndex = Mx.index;

    if (Mx.state[activeIndex] === undefined) {
      Mx.state[activeIndex] = initialState;
    }

    const currentState = Mx.state[activeIndex];

    const setter = (newValue: S) => {
      if (Mx.state[activeIndex] !== newValue) {
        Mx.state[activeIndex] = newValue;
        Mx.renderTree();
      }
      return newValue;
    };

    Mx.index++;

    return [currentState, setter];
  },
  render: (store, actions) => component => mountPoint => {
    const dispatchActions = Object.entries(actions).reduce((actions, [key, action]) => ({
      ...actions,
      [key]: (...args: any[]) => store.dispatch(action(...args))
    }), {});

    Mx.renderTree = () => {
      const currentTree = component({...Mx.currentState, ...dispatchActions});
      updateTree(mountPoint, currentTree, Mx.currentTree);

      Mx.currentTree = currentTree;
      Mx.index = 0;
    };

    Mx.currentState = store.getState();

    store.subscribe((state) => {
      Mx.currentState = state;
      Mx.renderTree();
    });

    return Mx.renderTree();
  }
};

export const useState = Mx.useState;
