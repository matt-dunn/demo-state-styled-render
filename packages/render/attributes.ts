/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { AnyFunc, HTMLElementMap, Props } from "./types";
import { propertyMap } from "./utils";

export const setAttribute = (
  element: HTMLElementMap,
  name: string,
  value: string | AnyFunc
) => {
  const propertyName = propertyMap(name);

  if (typeof value === "function") {
    element[propertyName] = value;
  } else {
    if (value) {
      element.setAttribute(propertyName, value);
    } else {
      element.removeAttribute(propertyName);
    }
    element[propertyName] = value;
  }
};

export const setAttributes = (element: HTMLElement, props: Props) =>
  Object.keys(props || {}).forEach((name) =>
    setAttribute(element, name, props[name])
  );

export const removeAttribute = (element: HTMLElement, name: string) =>
  element.removeAttribute(propertyMap(name));

export const updateAttribute = (
  element: HTMLElement,
  name: string,
  value: string,
  prevValue: string
) => {
  if (value !== prevValue) {
    if (value === undefined) {
      removeAttribute(element, name);
    } else if (!prevValue || value !== prevValue) {
      setAttribute(element, name, value);
    }
  }
};

export const updateAttributes = (
  element: HTMLElement,
  props: Props,
  prevProps: Props = {}
) =>
  Object.keys({ ...props, ...prevProps }).forEach((name) =>
    updateAttribute(element, name, props[name], prevProps[name])
  );
