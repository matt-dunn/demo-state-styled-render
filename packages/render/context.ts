/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { Node } from "packages/render";

import { looseRef } from "./utils";

export const getValueSymbol = Symbol("getValueSymbol");

export type Context<T = any> = {
  [getValueSymbol]: () => T | undefined;
  Provider: ({ value, children }: { value: T; children: Node }) => Node;
};

export const createContext = <T = any>(defaultValue?: T): Context<T> => {
  const currentValue = looseRef(defaultValue);

  return {
    [getValueSymbol]: () => currentValue.current,
    Provider: ({ value, children }) => {
      if (currentValue.current !== value) {
        currentValue.current = value;
      }

      return children;
    },
  };
};
