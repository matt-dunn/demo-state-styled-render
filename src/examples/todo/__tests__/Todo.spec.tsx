/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

/** @jsx jsx **/

import {jsx} from "packages/render";
import {mount} from "test/utils";

import {Todo} from "../Todo";

describe("Todo: Example of using mount", () => {
  it("should render component", async () => {
    const fixture = {id: "1", text:"My Todo Item", complete: true};

    const deleteTodo: any = jest.fn();
    const updateTodo: any = jest.fn();

    const wrapper = mount(
      <Todo todo={fixture} deleteTodo={deleteTodo} updateTodo={updateTodo} />
    );

    expect(wrapper.getDOMNode()).toMatchSnapshot();
  });
});
