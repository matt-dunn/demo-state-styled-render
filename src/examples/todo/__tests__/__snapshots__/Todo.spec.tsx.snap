// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo: Example of using mount should render component 1`] = `
<article
  class=" ms__enAiqr"
>
  <label>
    <del
      class="text-muted"
    >
      <input
        checked="true"
        type="checkbox"
      />
      My Todo Item
    </del>
  </label>
  <button
    class="btn btn-sm btn-outline-secondary"
  >
    Delete
  </button>
</article>
`;
